### ISSUE TITLE
Aard & BitTorrent

### ISSUE THREAD

**Issue by cosmopol** [itkach/aard2-android#47]
Firstly, the Aard databases download is really too error prone IMHO. I myself have wasted several days with trying getting the desired content for my purposes.

Secondly, direct downloads almost always fail for larger files due to tight traffic restrictions, even the tiny torrent files often are not accessible any more.

Thirdly, the torrent files that were available most often did not respect the <4GB file size limit, making it impossible to work with on most standard Android devices which are bound to the restrictions of the FAT32 file system format. One cannot even use the standard play store uTorrent app to download and seed the databases resulting in crashes, deadlocks, angered users, the need to have secondary devices for database retrieval and less seeds.

Fourthly, most often folks will be looking for the huge offline text database torrent downloads like Wikipedia on torrent sites, so releasing torrents to the Aard project website or even to personal storage accounts leaving the popular torrent sites out will have missed a great part of the community. Nobody automatically connects Aard with Wikipedia for instance. It would help folks to connect the dots.

Fifthly, therefore I am going to release selected Aard databases on mininova.org and legittorrents.info, which exclusively engage in the distribution of legal content, soon. Please keep me posted about stable releases of new torrent files that respect the <4GB file size limit if the need may be.

Sixthly, I would recommend to overwork the download section a litte bit with what I have stated above in mind.

Seventhly, did I miss a point here?

**Comment by itkach** [comment]
Please use http://aarddict.org/forum for this kind of posts. This is the wrong place for it: this is issue tracker for aard2-android bugs and features, and none of what you're talking about can be addressed by code changes in this repo. Also, there are a lot more people on the http://aarddict.org/forum mailing list than watching this repo. And there already are download issue discussions at http://aarddict.org/forum.

Now to your points

> direct downloads almost always fail for larger files due to tight traffic restrictions, even the tiny torrent files often are not accessible any more.

Indeed, cloud storage is ill suited for distribution of large popular files, although it's not a bad option for smaller and less popular ones, at least as an alternative/backup download. Torrent is the way to go, but ultimately it works well only if people use it and keep seeding. You can help by seeding.

> torrent files that were available most often did not respect the <4GB file size limit

Only English Wikipedia and recent German are bigger than 4Gb. @MHBraun who updates them (and does a great job!) distributes versions split into volumes <4Gb, so I'm not quite sure what's your complaint.

> making it impossible to work with on most standard Android devices which are bound to the restrictions of the FAT32 file system format.

Do not exaggerate. Internal storage on Android does not use FAT32 and doesn't have any issues with large files. Of those devices equipped with SD card slot, many do come with ExFAT support (Samsung, HTC) and have no trouble using SD cards larger than 32 Gb with large files. Devices that do not ship with ExFAT support (or support for any other usable filesystem for SD cards) are usually ones running vanilla Android and it's Google's fault, complain to them. Having said that, it is possible to split large dictionaries and, as said above, that's how English and German Wikipedias are compiled.

> One cannot even use the standard play store uTorrent app to download and seed the databases resulting in crashes, deadlocks, angered users

Sounds like uTorrent is buggy on Android. You are not seriously blaming aard for another application's defects?


> most often folks will be looking for the huge offline text database torrent downloads like Wikipedia on torrent sites

Says who?

> releasing torrents to the Aard project website or even to personal storage accounts leaving the popular torrent sites out will have missed a great part of the community. Nobody automatically connects Aard with Wikipedia for instance. It would help folks to connect the dots.

It would be silly if project website didn't have any pointers to downloading content, wouldn't it? If user found Aard 2 on the web or in Play Store hopefully they can read the description and follow the links. Not sure what "dots" need connecting so badly and how exactly do you propose to connect them.

> Fifthly, therefore I am going to release selected Aard databases on mininova.org and legittorrents.info, which exclusively engage in the distribution of legal content, soon. 

Sure, by all means, spread the word, good idea.

> Please keep me posted

Join the mailing list (http://aarddict.org/forum) and watch https://github.com/itkach/slob/wiki/Dictionaries
 
> Sixthly, I would recommend to overwork the download section a litte bit with what I have stated above in mind.

I'm not sure what do you mean by "download section". 

Download links for slob dictionaries are collected at  https://github.com/itkach/slob/wiki/Dictionaries which is a wiki page that any Github user can edit. Anyone can compile a dictionary, host files in any way that's available and/or convenient for them (torrent, cloud storage, own server) and share a link there. And then update and modify it as they see fit. Perhaps the page is a bit messy, but does the job, makes it easy to participate. Can you be more specific about improvements you suggest?

In any case, let me reiterate that forum is a much better place for this sort of discussions.

**Comment by francwalter** [comment]
French Wikipedia has grown now bigger than 4GB, but since now nobody had asked me to manufacture a split version yet.
Anyway there is not much download of it.
I think exFAT is really the default today for ext sd cards. I have FAT32 on my 64 gb sd but this is a special story with Cyanogen and absolut no default.

**Comment by cosmopol** [comment]
After having fought battles desperately for long, maybe 1 out of 10^4 users will embark on that epic journey to find the project's island in the vast ocean of the internet's projects, settle on that remote island, learn its inhabitant's unconventional language, reluctantly start to study island history to finally hike to your developers mountain hut and ask you to craft and hand out to them the sacred 4GB chunks of life...

Maybe not...



Do not take it personally. I have been using Aard for years now and it has become the most important Android app to me. I highly appreciate all your efforts. The math formulas look great. Kiwix only has images to offer here. Aard2 still only has some 1000+ downloads albeit all its benefits like lower storage space consumption. Aard1 has some 10000+ downloads. Given the fact that retrieval and installation of the databases is such a pain, how many common users do you expect to really use Aard regularily with the most recent database updates? Most newbies will give up with the first problem. Vanilla Android is the standard Google Android and the baseline for every project. Ignoring it is not a good idea (equals coming up with make-work for others :-) ). I personally ran into the 4GB limit issue with the extremely popular uTorrent Android app that has some 50 to 100 million downloads and a small project app called TorrentEx (after having downloaded for days each, of course, that is). (BTW: TorrentEx crashed on using the aard magnet links, too.) It really made me mad. Why not split every database in the first place and safe yourself all the trouble? Set the constants appropriately in the conversion tools and off you go.

One could incorporate torrentlib into Aard and maybe solve a lot of problems (maybe I will do just that in the future). mininova.org for instance does claim to webseed registered torrents for free. Everyone of us sets up a server (e.g. 'Raspberry Pi') additionally for his or hers home internet connection and off we go with some estimated 20 Mbit+ seed without counting in new peers if the need may be. (BTW: Some Aard databases are not available as torrent files on the project homepage rather on personal storage only.)

Type 'Wikipedia' into Google Play Store and you will not get Aard as hit. Nobody looks for Aard in the first place, only after having gotten to know it from studying articles like 'Wikipedia unterwegs' ( https://de.wikipedia.org/wiki/Wikipedia:Unterwegs ). There are several apps on the Play Store that have incorporated Wikipedia tags in their description and gotten some massive downloads. We have here this high quality product called Aard. Why not let other users get it to know and use seriously? The demand for it is out there.

**Comment by itkach** [comment]
> After having fought battles desperately for long, maybe 1 out of 10^4 users will embark on that epic journey to find the project's island in the vast ocean of the internet's projects, settle on that remote island, learn its inhabitant's unconventional language, reluctantly start to study island history to finally hike to your developers mountain hut and ask you to craft and hand out to them the sacred 4GB chunks of life...

How melodramatic

> Given the fact that retrieval and installation of the databases is such a pain, how many common users do you expect to really use Aard regularily with the most recent database updates?

I don't expect anything. Let me take this opportunity and explain that world dominance and huge number of users is not a goal of this project. I made it for me, because I wanted an app to lookup words from book reader apps and also it turned out to be a useful playground for me to learn and experiment. 
It is made available in the hope that it will be useful to others, and it's nice when people use and like your software, for sure, but that in itself is not a goal.

> Most newbies will give up with the first problem.

Fine with me. I believe there's a good deal of information available online, both about aarddict specifically and computing in general, and enough ways to ask for help that a lot of people can manage and figure it out. I don't mind helping out, but there's only so much energy I can put into it and hand-holding and babysitting newbies is not, generally, how I want to spend my time.

> I personally ran into the 4GB limit issue with the extremely popular uTorrent Android app
> TorrentEx crashed on using the aard magnet links

Again, sounds like these are defects in these apps. You make it sound like it's aard's responsibility to work around defects of other programs. I reject this notion. 

> One could incorporate torrentlib into Aard

I fail to see why Aard needs to become yet another torrent downloader. It does one thing - looks up words - and that's the way I like it.

> One could incorporate torrentlib into Aard and maybe solve a lot of problems (maybe I will do just that in the future). 

Sure, by all means - that's what's good about open source, you can take something and make it better.

> mininova.org for instance does claim to webseed registered torrents for free.

I'd be skeptical. Hosting and traffic costs money, so most likely this is subject to restrictions just like free personal cloud storage.

> Why not split every database in the first place and safe yourself all the trouble?

See, different people have different definitions of what constitutes "trouble". I don't know if I'd go as far as calling dealing with multiple files and splitting "trouble", but I also have no compelling reason to do it myself.

> There are several apps on the Play Store that have incorporated Wikipedia tags in their description and gotten some massive downloads

What are those magical "Wikipedia tags"? Is there some specific text you think should be there in the app description in Play store? I'm open to suggestions.

> Why not let other users get it to know and use seriously?

Let me try summarize what you're saying without the drama and exaggerations:

- *Small user base is a serious problem for Aard2*

It is not.

- *Dictionary downloads suck*. 

Agreed. None of what you propose is going to fix that though. Files under 4Gb are still too big and http downloads from cloud storage often fail. Torrents are good, but not everybody understands how to use them. Teaching aard2-android to download torrents might help with ease of use for some users, but I think you underestimate the complexity of building a proper torrent client. Doing so would be detrimental to the codebase, will introduce a whole slew of new issues while providing functionality that already is available through other applications. Not worth it. Torrents also tend to be ineffective if very few people seed which consistently has been the case with dictionary downloads. Also not everybody trusts torrents, some people seem to think that torrent technology itself if illegal. The only way to really fix downloads is to pay for hosting and traffic.

- *Files larger than 4Gb is a serious problem*

Disagreed. It was a serious problem for you and you jump to conclusion that it's a serious problem for everybody. Agreed that it is a problem for some users. A workaround is available for them.

- *Users don't know/can't find out about Aard2*

Agreed. I'm open to suggestions on how to fix that.

**Comment by francwalter** [comment]
> ...It is made available in the hope that it will be useful to others...
Hope in my case fulfilled :) Thank!!!

**Comment by MHBraun** [comment]
Igor,
please close this discussion. It does not belong to here. Far better place is the forum 
http://aarddict.org/forum 
as you stated already.
Furthermore there is no constructive support or solution provided. This does not help in any way.
We may complain how bad the world is, as long as we do not do something about it, it will not change.

**Comment by cosmopol** [comment]
I see. All that improvements only make sense if a broader user base will be the project's goal. I will always regard Mr Tkachs advice before mine for I do not consider myself an IT pro. You should really come up with some kind of manifesto, roadmap etc. for new contributors to the project. Emmanuel Engelhart did so with his Kiwix project. Although I consider Aard technically superior for my use case, contributing to the Kiwix project appeals more to me. This is my last post for the time being.

Happy new year!

**Comment by francwalter** [comment]
Happy new Year!!
All togehter you!
Thank you for the contribution to have such a great tool in handy :) 

@cosmopol good bye and I wish you a good welcome in kiwix contribution! I went the other way, from kiwix to aard.

**Comment by cosmopol** [comment]
@francwalter May I ask what your motivation was/is and your experiences with contribution to the two projects Kiwix and Aard were/are?

**Comment by cosmopol** [comment]
@francwalter May I ask what your motivation was/is and your experiences with contribution to the two projects Kiwix and Aard were/are?

(please post on the forum http://aarddict.org/forum/ / We could just delete the non-repository text entries later on here at GitHub)

**Comment by francwalter** [comment]
@cosmopol : I built the wiki for kiwix and had some difficulties with the mwoffliner.js, that took me long therefore. And there was nearly no help from the developer (Engelhardt), it was quite different than this project here is.
At this time I thought that kiwix was not ment to contribute in using mwoffliner.js and related code. Maybe this is different now, but I was a little fed up with all this.

And because I don't need offline pictures (the really only advantage of kiwix, if needed) because this means some 20 GB for lousy small pictures, and I found aard2, I quit this project kiwix. 
The Android app was at this time not good neither.

As this issue is closed anyway, I could post these two posts here better than in aard google groups, where kiwix not matters at all :)

**Comment by cosmopol** [comment]
@francwalter : OK, I hope we will not be kicked for posting the "wrong" texts.

I am not familiar with the issues surrounding the mwoffliner.js code (yet).

I always was enthusiastic about the Wikimedia projects, particularily Wikipedia. I feel like contributing to a meaningful project in my free time, let's say as a hobby. I have been working as IT administrator / "programmer light" for some years and have some understanding of what is going on here. Aard might be technically superior, but Kiwix is actually meant to be used by a broad user base by design. Cross-platform (although Android alone should be the best bet for the future), all languages, instant downloads and torrenting, a complete experience in many ways. It even is used for the OLPC project whereas Aard is rather used by some semi-pros only.

The offline pictues might come in handy when you deal with abstract articles, like often found in math, other than that tiny pictures do not work well. Anyhow, in the near future offline pictures will be the standard.

I always liked the idea of being able to have all wikis on me all the time and read in them daily. Moreover projects like Kiwix or Aard are potentially able to deliver the future standard works of knowledge to everybody everywhere. Consider how many people even today are virtually offline.

**Comment by francwalter** [comment]
@cosmopol : maybe I am to strong with kiwix and now it is different, I didnt check for nearly exact 2 years, as I just checked in my notes. After reading these notes, I remember a bit how difficult (for me) all this was. That was far away from beeing straight forward, this downloading of wikipedia with that mwoffliner.js (that was the correspondent to mwscrape in aard). Once E.Engelhardt told me that it was not yet ment to be public, that why he didnt provide any support. It was (?) his own little baby, despite beeing opensource.

Your right, once the offline wiki will have pictures as default, I guess too. Space will be much less problem. But could be also that offline will be so rare that all will be online and offline wiki would go down, also I could imagine this scenario. In the rich world only -  like you pictured :(
But Math pictures in aard2 are no pictures and work completely offline as I know.

I am like you admin IT and only light programer from all a little bit :)

How is the kiwix wikis up to date nowadays? It was not good 2 years ago. That why I compiled a dewiki myself, because only old dewiki for download. But the app was not good neither.
Did you test that all?
